### ISSUE TITLE
Cronicle crashing while running a job

### ISSUE THREAD

**Issue by zaidoonhad** [jhuckaby/Cronicle#81]
## Summary

Cronicle is crashing suddenly while running a job.

## Steps to reproduce the problem

I started a job, and after 18 hours, Cronicle crashed. This was never an issue before, and nothing has changed. I didn't update/modify/remove anything.

## Your Setup
I installed Cronicle on my private server and ran it.

### Operating system and version?
14.04.5 LTS

### Node.js version?
10.4.1

### Cronicle software version?
0.8.6

### Are you using a multi-server setup, or just a single server?
Single
### Are you using the filesystem as back-end storage, or S3/Couchbase?
filesystem

### Can you reproduce the crash consistently?
no, this is the first time, and so far, it didn't happen again.
## Log Excerpts

Fri Jul 20 2018 09:35:57 GMT+0000 (Greenwich Mean Time)
TypeError [ERR_INVALID_ARG_VALUE]: The argument 'path' must be a string or Uint8Array without null bytes. Received '/opt/cronicle/htdocs/web-console/ServerInfo.jsp\u0000'
    at Object.fs.stat (fs.js:918:3)
    at Server.servePath (/opt/cronicle/node_modules/node-static/lib/node-static.js:142:12)
    at /opt/cronicle/node_modules/node-static/lib/node-static.js:183:14
    at process._tickCallback (internal/process/next_tick.js:61:11)

**Comment by jhuckaby** [comment]
Oh my, thank you very much for catching this.  Looks like a random attack by a bot looking for Java/JSP vulnerabilities.  There is no JSP code anywhere in Cronicle, so this request came from "outside" (are you running Cronicle on a public facing server?).

I will get a fix out ASAP!  Thanks again!

**Comment by zaidoonhad** [comment]
Thank you for the prompt response and great tool. I'm actually using this
on a server that can only be accessed internally within my company. This
probably happened because one of our security tools were scanning the
server for vulnerabilities. Thanks again.

**Comment by jhuckaby** [comment]
Should be fixed in [Cronicle v0.8.16](https://github.com/jhuckaby/Cronicle/releases/tag/v0.8.16).  Thanks again!

**Comment by zaidoonhad** [comment]
Cronicle is still crashing due to this issue. I am on the latest version of Cronicle and nothing else has changed:

Tue Aug 07 2018 15:57:41 GMT+0000 (Greenwich Mean Time)
TypeError [ERR_INVALID_ARG_VALUE]: The argument 'path' must be a string or Uint8Array without null bytes. Received '/opt/cronicle/htdocs/web-console/ServerInfo.jsp\u0000'
    at Object.fs.stat (fs.js:918:3)
    at Server.servePath (/opt/cronicle/node_modules/node-static/lib/node-static.js:142:12)
    at /opt/cronicle/node_modules/node-static/lib/node-static.js:183:14
    at process._tickCallback (internal/process/next_tick.js:61:11)

Thank you for your help.

**Comment by jhuckaby** [comment]
Please note that the bug was actually in a dependency module ([pixl-server-web](https://www.npmjs.com/package/pixl-server-web)), i.e. not Cronicle itself.  So if you performed some kind of manual update to get Cronicle up to v0.8.16, and didn't also update the npm dependencies, it wouldn't have fixed the bug.  You may have to update the npm dependencies by hand on your server:

```
cd /opt/cronicle
sudo npm install
sudo bin/control.sh restart
```

However, if you used the standard Cronicle upgrade instructions and upgrade script, then it should have also updated all dependencies, and I cannot explain how you are still experiencing the bug.  If you could, please tell me what version of pixl-server-web you have on your Cronicle server where the crash is occurring:

```
grep version /opt/cronicle/node_modules/pixl-server-web/package.json
```

Thank you.

**Comment by zaidoonhad** [comment]
I used control.sh upgrade to upgrade to the latest version of Cronicle and my pixel-server-web version is:
grep version /opt/cronicle/node_modules/pixl-server-web/package.json 
    "type": "version",
  "version": "1.1.5"

**Comment by jhuckaby** [comment]
That is really bizarre.  It really should be fixed in that version.  I guess I will re-open the issue, but I cannot fathom how it is crashing at that same place in the code.  I'll look at this again as soon as I have time.  Sorry for the inconvenience.

**Comment by zaidoonhad** [comment]
Here is the output of gzip -cd /opt/cronicle/logs/archives/2018/08/*/WebServer* | grep ServerInfo.jsp:
[1533657461.847][2018-08-07 15:57:41][WebServer][debug][8][New HTTP request: GET /web-console/ServerInfo.jsp (::ffff:xxx.xx.xxx.xxx)][{"socket":"c1238","version":"1.1"}]
[1533657461.847][2018-08-07 15:57:41][WebServer][debug][9][Serving static file for: /web-console/ServerInfo.jsp][]
[1533657461.848][2018-08-07 15:57:41][WebServer][error][404][Error serving static file: /web-console/ServerInfo.jsp: HTTP 404 Not Found][{"ips":["::ffff:xxx.xx.xxx.xxx"],"useragent":"Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)","referrer":"","cookie":"","url":"http://hostname:3012/web-console/ServerInfo.jsp"}]

Thank you.

**Comment by zaidoonhad** [comment]
it crashed again this morning at the same place in the code, while I was running a job, however, even though Cronicle said  Error: Aborted Job: Server 'xxx' shut down unexpectedly. The job is still running in the background (I ran the job using the shell script). Would it be possible to terminate the running jobs when the server is shutdown unexpectedly?

**Comment by zaidoonhad** [comment]
I just read this part:

"When a master server experiences a catastrophic failure such as a daemon crash, kernel panic or power loss, it has no time to do anything, so any active jobs on the server are instantly dead. The jobs will eventually be logged as failures, and the logs recovered when the server comes back online."

however, that is not happening for me, any suggestions? Thank you.

**Comment by jhuckaby** [comment]
Thank you for that log excerpt.  Curiously, that particular request doesn't look like it ended in a crash.  The URL looks clean, has no weird character at the end, and Cronicle sent back a correct HTTP 404 response.

However, I THINK I see where the crash may be occurring.  This is just a wild shot in the dark, but if a URL comes in with an *encoded* ASCII NULL in the form of `%00` then it seems to reproduce the crash.  I will release a patch that covers this exact case, even though I don't see what I am expecting to see in your logs.

Yup, the master server crash not aborting jobs (specifically non-detached jobs) is a separate bug.  I will add that to my list as well.  Thank you!

I'll leave this issue open and flagged until both things are fixed.

**Comment by jhuckaby** [comment]
Both issues should be completely fixed in [Version 0.8.20](https://github.com/jhuckaby/Cronicle/releases/tag/v0.8.20).  Thanks again for your help!

**Comment by zaidoonhad** [comment]
thank you
